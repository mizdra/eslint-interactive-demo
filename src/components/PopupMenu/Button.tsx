import { ComponentChildren } from 'preact';

export type PopupMenuButtonProps = {
  selected?: boolean;
  children: ComponentChildren;
};

export function PopupMenuButton({ children, selected }: PopupMenuButtonProps) {
  return <div className={selected ? 'button selected' : 'button'}>{children}</div>;
}
