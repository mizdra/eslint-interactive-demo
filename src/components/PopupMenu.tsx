import { useCallback, useEffect, useMemo, useState } from 'preact/hooks';
import useResizeObserver from 'use-resize-observer';
import { useDocumentEventListener } from '../hooks/useDocumentEventListener';
import { useScrapbox } from '../hooks/useScrapbox';
import { calcButtonContainerStyle, calcPopupMenuStyle, calcTriangleStyle } from '../lib/calc-style';
import { Icon } from '../lib/icon';
import { isComposing } from '../lib/key';
import { CursorPosition } from '../types';
import { PopupMenuButton } from './PopupMenu/Button';

const DEFAULT_IS_POPUP_CLOSE_KEY_DOWN = (e: KeyboardEvent) => {
  return e.key === 'Escape' && !e.ctrlKey && !e.shiftKey && !e.altKey;
};

export type PopupMenuProps = {
  open: boolean;
  emptyMessage?: string;
  cursorPosition: CursorPosition;
  icons: Icon[];
  onSelect?: (icon: Icon, index: number) => void;
  onClose?: () => void;
  isPopupCloseKeyDown?: (e: KeyboardEvent) => boolean;
};

export function PopupMenu({
  open,
  emptyMessage,
  cursorPosition,
  icons,
  onSelect,
  onClose,
  isPopupCloseKeyDown = DEFAULT_IS_POPUP_CLOSE_KEY_DOWN,
}: PopupMenuProps) {
  const { editor } = useScrapbox();
  const { ref, width: buttonContainerWidth = 0 } = useResizeObserver<HTMLDivElement>();
  const isEmpty = useMemo(() => icons.length === 0, [icons.length]);

  const [selectedIndex, setSelectedIndex] = useState<number | null>(null);
  const { width: editorWidth = 0 } = useResizeObserver({ ref: editor });

  // icons が変わったら選択位置を 0 番目に戻す。ただし空なら null にセットする。
  useEffect(() => {
    setSelectedIndex(isEmpty ? null : 0);
  }, [isEmpty, icons]);

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      // 閉じている時は何もしない
      if (!open) return;
      // IMEによる変換中は何もしない
      if (isComposing(e)) return;

      const isTab = e.key === 'Tab' && !e.ctrlKey && !e.shiftKey && !e.altKey;
      const isShiftTab = e.key === 'Tab' && !e.ctrlKey && e.shiftKey && !e.altKey;
      const isEnter = e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && !e.altKey;
      const isClose = isPopupCloseKeyDown(e);

      if (isTab || isShiftTab || isEnter || isClose) {
        e.preventDefault();
        e.stopPropagation();
      }

      if (isEmpty || selectedIndex === null) {
        if (isClose) onClose?.();
      } else {
        if (isTab) setSelectedIndex((selectedIndex + 1) % icons.length);
        if (isShiftTab) setSelectedIndex((selectedIndex - 1 + icons.length) % icons.length);
        if (isEnter) onSelect?.(icons[selectedIndex], selectedIndex);
        if (isClose) onClose?.();
      }
    },
    [isEmpty, isPopupCloseKeyDown, icons, onClose, onSelect, open, selectedIndex],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  const popupMenuStyle = calcPopupMenuStyle(cursorPosition);
  const triangleStyle = calcTriangleStyle(cursorPosition, isEmpty);
  const buttonContainerStyle = calcButtonContainerStyle(editorWidth, buttonContainerWidth, cursorPosition, isEmpty);

  const iconListElement = icons.map((icon, i) => (
    <PopupMenuButton key={icon.key} selected={selectedIndex === i}>
      {icon.element}
    </PopupMenuButton>
  ));

  return (
    <>
      {open && (
        <div className="popup-menu" style={popupMenuStyle} data-testid="popup-menu">
          <div ref={ref} className="button-container" style={buttonContainerStyle} data-testid="button-container">
            {icons.length === 0 ? emptyMessage ?? 'アイテムは空です' : iconListElement}
          </div>
          <div className="triangle" style={triangleStyle} />
        </div>
      )}
    </>
  );
}
