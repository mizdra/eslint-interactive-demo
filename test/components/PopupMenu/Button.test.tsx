import { render } from '@testing-library/preact';
import { PopupMenuButton } from '../../../src/components/PopupMenu/Button';

describe('Button', () => {
  test('children が表示される', () => {
    const { container } = render(<PopupMenuButton>text</PopupMenuButton>);
    expect(container).toMatchSnapshot();
  });
  test('selected === true の時、選択状態であることを表すクラスが付与される', () => {
    const { container } = render(<PopupMenuButton selected>text</PopupMenuButton>);
    expect(container).toMatchSnapshot();
  });
});
