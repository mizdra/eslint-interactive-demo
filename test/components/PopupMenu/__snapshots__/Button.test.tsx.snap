// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button children が表示される 1`] = `
<div>
  <div
    class="button"
  >
    text
  </div>
</div>
`;

exports[`Button selected === true の時、選択状態であることを表すクラスが付与される 1`] = `
<div>
  <div
    class="button selected"
  >
    text
  </div>
</div>
`;
